binder/src/test/java/io/grpc/binder/BinderTransportSecurityTest.java

```diff
@@ -0,0 +1,235 @@
+    shadowOf(Looper.getMainLooper()).idle();
+  }
+
+  private ListenableFuture<Status> makeCall() {
+    ClientCall<Empty, Empty> call =
+        channel.newCall(
+            getMethodDescriptor(),
+            CallOptions.DEFAULT.withExecutor(service.getExecutor()));
+    ListenableFuture<Empty> responseFuture =
+        ClientCalls.futureUnaryCall(call, Empty.getDefaultInstance());
+
+    idleLoopers();
+
+    return Futures.catching(
+        Futures.transform(responseFuture, unused -> Status.OK, directExecutor()),
+        StatusRuntimeException.class,
+        StatusRuntimeException::getStatus,
+        directExecutor());
+  }
+
+  private static MethodDescriptor<Empty, Empty> getMethodDescriptor() {
+    MethodDescriptor.Marshaller<Empty> marshaller =
+        ProtoLiteUtils.marshaller(Empty.getDefaultInstance());
+
+    return MethodDescriptor.newBuilder(marshaller, marshaller)
+        .setFullMethodName(FULL_METHOD_NAME)
+        .setType(MethodDescriptor.MethodType.UNARY)
+        .setSampledToLocalTracing(true)
+        .build();
+  }
+
+  private static class SomeService extends LifecycleService {
+
+    private final IBinderReceiver binderReceiver = new IBinderReceiver();
+    private final ArrayBlockingQueue<SettableFuture<Status>> statusesToSet =
+        new ArrayBlockingQueue<>(128);
+    private Server server;
+    private HandlerThread handlerThread;
+    private Handler handler;
+
+    @Override
+    public void onCreate() {
+      super.onCreate();
+      handlerThread = new HandlerThread("test_handler_thread");
+      handlerThread.start();
+      handler = new Handler(handlerThread.getLooper());
+
+      MethodDescriptor<Empty, Empty> methodDesc = getMethodDescriptor();
+      ServerCallHandler<Empty, Empty> callHandler =
+          ServerCalls.asyncUnaryCall(
+              (req, respObserver) -> {
+                respObserver.onNext(req);
+                respObserver.onCompleted();
+              });
+      ServerMethodDefinition<Empty, Empty> methodDef =
+          ServerMethodDefinition.create(methodDesc, callHandler);
+      ServerServiceDefinition def = ServerServiceDefinition.builder(SERVICE_NAME)
+          .addMethod(methodDef)
+          .build();
+
+      server = BinderServerBuilder.forAddress(
+              AndroidComponentAddress.forContext(this),
+              binderReceiver)
+          .addService(def)
+          .securityPolicy(ServerSecurityPolicy.newBuilder()
+              .servicePolicy(SERVICE_NAME, new AsyncSecurityPolicy() {
+                @Override
+                ListenableFuture<Status> checkAuthorizationAsync(int uid) {
+                  return Futures.submitAsync(() -> {
+                    SettableFuture<Status> status = SettableFuture.create();
+                    statusesToSet.add(status);
+                    return status;
+                  }, getExecutor());
+                }
+              })
+              .build())
+          .executor(getExecutor())
```

Comment:
I worry this is going to be flaky unless you also idle the scheduledExecutorService().

Reply:
It's harder to get a `ScheduledExecutorService` that we can wait to go idle than an `Executor`. Still, tried to implement a minimally working version. Passing it here and elsewhere.